### ISSUE TITLE
Switch Channel from 1.19/stable to 1.20/stable fails for HA cluster

### ISSUE THREAD

**Issue by arctiqaurora** [ubuntu/microk8s#1877]
Please run `microk8s inspect` and attach the generated tarball to this issue.

We appreciate your feedback. Thank you for using microk8s.

I have an HA mickok8s cluster on channel 1.19/stable with 4 nodes. When trying to switch channel to 1.20/stable, it failed and this is the error message:
```
ubuntu@kubernetes0:~$ sudo snap refresh microk8s --classic --channel 1.20/stable
Consider re-refresh of "microk8s"                                                                                                                                                           \
error: cannot perform the following tasks:
- Run configure hook of "microk8s" snap if present (run hook "configure": 
-----
--requestheader-group-headers=X-Remote-Group
--requestheader-username-headers=X-Remote-User
--proxy-client-cert-file=${SNAP_DATA}/certs/front-proxy-client.crt
--proxy-client-key-file=${SNAP_DATA}/certs/front-proxy-client.key
+ grep -e requestheader-client-ca-file=/var/snap/microk8s/.../certs/ca.crt /var/snap/microk8s/1864/args/kube-apiserver
+ grep -F 'requestheader-client-ca-file=${SNAP_DATA}/certs/ca.crt' /var/snap/microk8s/1864/args/kube-apiserver
+ mkdir -p /var/snap/microk8s/1864/var/lock/
+ '[' '!' -f /var/snap/microk8s/1864/credentials/client.config ']'
+ '[' '!' -e /var/snap/microk8s/1864/args/kubectl-env ']'
+ '[' '!' -e /var/snap/microk8s/1864/args/kubectl ']'
+ '[' '!' -e /var/snap/microk8s/1864/args/containerd ']'
+ grep -e '\-\-docker unix://${SNAP_DATA}/docker.sock' /var/snap/microk8s/1864/args/kubelet
++ produce_certs
++ export OPENSSL_CONF=/snap/microk8s/current/etc/ssl/openssl.cnf
++ OPENSSL_CONF=/snap/microk8s/current/etc/ssl/openssl.cnf
++ for key in serviceaccount.key ca.key server.key front-proxy-ca.key front-proxy-client.key
++ '[' -f /var/snap/microk8s/1864/certs/serviceaccount.key ']'
++ for key in serviceaccount.key ca.key server.key front-proxy-ca.key front-proxy-client.key
++ '[' -f /var/snap/microk8s/1864/certs/ca.key ']'
++ for key in serviceaccount.key ca.key server.key front-proxy-ca.key front-proxy-client.key
++ '[' -f /var/snap/microk8s/1864/certs/server.key ']'
++ for key in serviceaccount.key ca.key server.key front-proxy-ca.key front-proxy-client.key
++ '[' -f /var/snap/microk8s/1864/certs/front-proxy-ca.key ']'
++ for key in serviceaccount.key ca.key server.key front-proxy-ca.key front-proxy-client.key
++ '[' -f /var/snap/microk8s/1864/certs/front-proxy-client.key ']'
++ '[' -f /var/snap/microk8s/1864/certs/ca.crt ']'
++ '[' -f /var/snap/microk8s/1864/certs/front-proxy-ca.crt ']'
++ render_csr_conf
+++ get_ips
++++ /snap/microk8s/1864/bin/hostname -I
+++ local 'IP_ADDR=10.100.200.120 10.1.206.0 '
+++ [[ -z 10.100.200.120 10.1.206.0  ]]
+++ /snap/microk8s/1864/sbin/ifconfig cni0
+++ echo '10.100.200.120 10.1.206.0 '
++ local 'IP_ADDRESSES=10.100.200.120 10.1.206.0 '
++ cp /var/snap/microk8s/1864/certs/csr.conf.template /var/snap/microk8s/1864/certs/csr.conf.rendered
++ '[' '10.100.200.120 10.1.206.0 ' == 127.0.0.1 ']'
++ '[' '10.100.200.120 10.1.206.0 ' == none ']'
++ local ips= sep=
++ local -i i=3
+++ echo '10.100.200.120 10.1.206.0 '
++ for IP_ADDR in $(echo "$IP_ADDRESSES")
++ ips+='IP.3 = 10.100.200.120'
++ sep='\n'
++ for IP_ADDR in $(echo "$IP_ADDRESSES")
++ ips+='\nIP.4 = 10.1.206.0'
++ sep='\n'
++ /snap/microk8s/1864/bin/sed -i 's/#MOREIPS/IP.3 = 10.100.200.120\nIP.4 = 10.1.206.0/g' /var/snap/microk8s/1864/certs/csr.conf.rendered
++ '[' -f /var/snap/microk8s/1864/certs/csr.conf ']'
++ local force
++ /snap/microk8s/1864/usr/bin/cmp -s /var/snap/microk8s/1864/certs/csr.conf.rendered /var/snap/microk8s/1864/certs/csr.conf
++ force=false
++ false
++ '[' '!' -f /var/snap/microk8s/1864/certs/front-proxy-client.crt ']'
+++ /snap/microk8s/1864/usr/bin/openssl x509 -noout -issuer
++ '[' 'issuer= /CN=front-proxy-ca' == 'issuer=CN = 127.0.0.1' ']'
++ echo 0
+ '[' 0 == 1 ']'
+ '[' -e /var/snap/microk8s/1864/args/containerd-template.toml ']'
+ grep -e 'stream_server_address = ""' /var/snap/microk8s/1864/args/containerd-template.toml
+ grep -e '\-\-allow-privileged' /var/snap/microk8s/1864/args/kubelet
+ '[' -f /root/snap/microk8s/common/istio-auth.lock ']'
+ '[' -f /root/snap/microk8s/common/istio-auth.lock ']'
+ need_api_restart=false
+ '[' -f /var/snap/microk8s/1864/credentials/kubelet.config ']'
+ '[' -f /var/snap/microk8s/1864/credentials/proxy.config ']'
+ '[' -f /var/snap/microk8s/1864/credentials/scheduler.config ']'
+ '[' -f /var/snap/microk8s/1864/credentials/controller.config ']'
+ for dir in "${SNAP_DATA}/credentials/ ${SNAP_DATA}/certs/ ${SNAP_DATA}/args/ ${SNAP_DATA}/var/lock"
+ chmod -R ug+rwX /var/snap/microk8s/1864/credentials/ /var/snap/microk8s/1864/certs/ /var/snap/microk8s/1864/args/ /var/snap/microk8s/1864/var/lock
+ chmod -R o-rwX /var/snap/microk8s/1864/credentials/ /var/snap/microk8s/1864/certs/ /var/snap/microk8s/1864/args/ /var/snap/microk8s/1864/var/lock
+ getent group microk8s
+ getent group microk8s
+ chgrp microk8s -R /var/snap/microk8s/1864/credentials/ /var/snap/microk8s/1864/certs/ /var/snap/microk8s/1864/args/ /var/snap/microk8s/1864/var/lock/ /var/snap/microk8s/1864/var/kubernetes/backend/
+ '[' -e /var/snap/microk8s/1864/opt/cni/bin/flanneld ']'
+ '[' -f /var/snap/microk8s/1864/args/flanneld ']'
+ grep -e etcd.socket:2379 /var/snap/microk8s/1864/args/etcd
+ grep -e basic-auth-file /var/snap/microk8s/1864/args/kube-apiserver
+ grep -e service-account-issuer /var/snap/microk8s/1864/args/kube-apiserver
+ refresh_opt_in_config service-account-issuer https://kubernetes.default.svc kube-apiserver
+ local opt=--service-account-issuer
+ local value=https://kubernetes.default.svc
+ local config_file=/var/snap/microk8s/1864/args/kube-apiserver
+ local replace_line=--service-account-issuer=https://kubernetes.default.svc
++ grep -qE '^--service-account-issuer=' /var/snap/microk8s/1864/args/kube-apiserver
+ run_with_sudo /snap/microk8s/1864/bin/sed -i '$ a --service-account-issuer=https://kubernetes.default.svc' /var/snap/microk8s/1864/args/kube-apiserver
+ '[' -n '' ']'
+ '[' /snap/microk8s/1864/bin/sed == preserve_env ']'
+ sudo /snap/microk8s/1864/bin/sed -i '$ a --service-account-issuer=https://kubernetes.default.svc' /var/snap/microk8s/1864/args/kube-apiserver
+ '[' -e /var/snap/microk8s/1864/var/lock/ha-cluster ']'
+ run_with_sudo preserve_env /snap/microk8s/1864/usr/bin/python3 /snap/microk8s/1864/scripts/cluster/distributed_op.py update_argument kube-apiserver --service-account-issuer https://kubernetes.default.svc
+ '[' -n '' ']'
+ '[' preserve_env == preserve_env ']'
+ shift
+ sudo -E /snap/microk8s/1864/usr/bin/python3 /snap/microk8s/1864/scripts/cluster/distributed_op.py update_argument kube-apiserver --service-account-issuer https://kubernetes.default.svc
The connection to the server 127.0.0.1:16443 was refused - did you specify the right host or port?
Adding argument --service-account-issuer to nodes.
Could not query for nodes
Command '['/snap/microk8s/1864/microk8s-kubectl.wrapper', 'get', 'no', '-o', 'json']' returned non-zero exit status 1
-----)
```
The node will stay in 1.19/stable. Here are the inspect reports before and after

**Comment by arctiqaurora** [comment]
[after.tar.gz](https://github.com/ubuntu/microk8s/files/5784277/after.tar.gz)
[before.tar.gz](https://github.com/ubuntu/microk8s/files/5784278/before.tar.gz)

**Comment by MichaelGooden** [comment]
We were experiencing this same problem moving from `1.19`, manually adding the follow lines to the `/var/snap/microk8s/args/kube-apiserver` args file allowed the remainder of the upgrade process to complete to `1.20/stable`.

```
--service-account-issuer='https://kubernetes.default.svc'
--service-account-signing-key-file=${SNAP_DATA}/certs/serviceaccount.key
```

**Comment by ktsakalozos** [comment]
Thank you @arctiqaurora for reporting this and @MichaelGooden for offering a workaround.

You are right this is a bug in the way we attempt to update nodes during an upgrade with the required arguments this is why you get the `Could not query for nodes`  error. Working on a Fix right now.

Thank you for  your patience.

**Comment by ktsakalozos** [comment]
The fix works for me. It is going to be available within the day on 1.20/egde. It would be great if you could give it a try. I will push it to 1.20/stable early next week. Thank you.

**Comment by arctiqaurora** [comment]
yes it worked for me
